File: Advanced Research Methodology Fall 2026/code/appB_tests.py
"""appB_tests.py -- Advanced Research Methodology, Appendix B.

Python versions of the procedures in the handout, for those who prefer it.
Each block generates its own illustrative data, so the file runs end to end.

Install once:
    pip install numpy pandas scipy statsmodels pingouin scikit-learn

Every number produced here is from simulated data. Nothing in this file is
a research finding.
"""

import numpy as np
import pandas as pd
import pingouin as pg
import statsmodels.formula.api as smf
from statsmodels.stats.multitest import multipletests
from statsmodels.stats.power import TTestIndPower

RNG = np.random.default_rng(20260902)  # every analysis fixes its seed


def two_groups():
    """Two independent groups: Welch, effect size, bootstrap interval."""
    a = RNG.normal(6.9, 2.5, 60)
    b = RNG.normal(6.0, 2.5, 60)

    # Welch by default -- does not assume equal variances.
    print(pg.ttest(a, b, correction=True))
    print("Cohen's d:", pg.compute_effsize(a, b, eftype="cohen"))

    # Distribution-free alternative.
    print(pg.mwu(a, b))

    # Bootstrap interval: no distributional assumption at all.
    diffs = [RNG.choice(a, 60).mean() - RNG.choice(b, 60).mean()
             for _ in range(5000)]
    print("Bootstrap 95% CI:", np.percentile(diffs, [2.5, 97.5]))


def nested():
    """Nested data. Ignoring the structure inflates significance."""
    n_dev, n_art = 34, 4
    rows = []
    dev_eff = RNG.normal(0, 1.4, n_dev)
    art_eff = RNG.normal(0, 0.8, n_art)
    for d in range(n_dev):
        for a in range(n_art):
            assistant = (d + a) % 2
            rows.append(dict(
                dev=d, artefact=a, assistant=assistant,
                found=6 + 0.9 * assistant + dev_eff[d] + art_eff[a]
                      + RNG.normal(0, 1.6)))
    df = pd.DataFrame(rows)

    naive = smf.ols("found ~ assistant", data=df).fit()
    mixed = smf.mixedlm("found ~ assistant", data=df, groups=df["dev"]).fit()

    print("Naive SE:", naive.bse["assistant"])
    print("Mixed SE:", mixed.bse["assistant"])
    print("-- the ratio is how much the naive model overstated precision.")


def counts():
    """Software counts are overdispersed. Poisson understates the error."""
    size = RNG.poisson(40, 300)
    defects = RNG.negative_binomial(1.2, 1.2 / (1.2 + 0.05 * size))
    df = pd.DataFrame({"size": size, "defects": defects})

    pois = smf.poisson("defects ~ size", data=df).fit(disp=0)
    negb = smf.negativebinomial("defects ~ size", data=df).fit(disp=0)

    print("Poisson SE:", pois.bse["size"], " NegBin SE:", negb.bse["size"])
    print("Poisson intervals are too narrow when variance exceeds the mean.")


def power_and_multiplicity():
    """A-priori power for the SESOI, and correction for declared tests."""
    n = TTestIndPower().solve_power(effect_size=0.40, alpha=0.05, power=0.80)
    print(f"Required n per group for d=0.40: {np.ceil(n):.0f}")

    pvals = [0.011, 0.031, 0.042, 0.180, 0.560]
    print("holm:", multipletests(pvals, method="holm")[1].round(4))
    print("BH:  ", multipletests(pvals, method="fdr_bh")[1].round(4))


def leakage_demo():
    """The leakage that matters most in software data: grouped splitting.

    Random row-level splitting lets a model memorise per-group base rates.
    The same model, honestly validated, performs far worse -- and that drop
    is the finding, not a failure (see Chapter 22).
    """
    from sklearn.ensemble import RandomForestClassifier
    from sklearn.model_selection import cross_val_score, GroupKFold, KFold

    n_groups, per_group = 20, 60
    groups = np.repeat(np.arange(n_groups), per_group)
    base = RNG.uniform(0.05, 0.6, n_groups)[groups]   # group-specific rate
    X = np.column_stack([RNG.normal(size=len(groups)),
                         base + RNG.normal(0, 0.05, len(groups))])
    y = RNG.binomial(1, base)

    clf = RandomForestClassifier(n_estimators=200, random_state=0)
    naive = cross_val_score(clf, X, y, cv=KFold(5, shuffle=True,
                                                random_state=0), scoring="roc_auc")
    honest = cross_val_score(clf, X, y, groups=groups,
                             cv=GroupKFold(5), scoring="roc_auc")

    print(f"Random  k-fold AUC: {naive.mean():.3f}  <- leaks group identity")
    print(f"Grouped k-fold AUC: {honest.mean():.3f}  <- honest")


if __name__ == "__main__":
    for name, fn in [("TWO GROUPS", two_groups),
                     ("NESTED DATA", nested),
                     ("COUNTS", counts),
                     ("POWER AND MULTIPLICITY", power_and_multiplicity),
                     ("LEAKAGE", leakage_demo)]:
        print(f"\n{'=' * 60}\n{name}\n{'=' * 60}")
        fn()
    print("\nAll data simulated. See Appendix B of the handout.")

File: Advanced Research Methodology Fall 2026/code/appB_tests.R
# =====================================================================
#  appB_tests.R -- Advanced Research Methodology, Appendix B
#
#  Runnable versions of the statistical procedures in the handout.
#  Each block is self-contained and generates its own illustrative data,
#  so the file runs end to end on a clean machine.
#
#  Install once:
#    install.packages(c("effsize", "lme4", "lmerTest", "MASS", "pwr",
#                       "TOSTER", "psych", "boot", "PMCMRplus"))
#
#  Every number produced here is from simulated data. Nothing in this
#  file is a research finding.
# =====================================================================

set.seed(20260902)   # every analysis in this course fixes its seed

# ---------------------------------------------------------------------
# 1. Two independent groups
#    Welch's t-test is R's default and should stay that way: it does not
#    assume equal variances and costs almost nothing when they are equal.
# ---------------------------------------------------------------------
d <- data.frame(
  g = rep(c("checklist", "unguided"), each = 60),
  y = c(rnorm(60, mean = 6.9, sd = 2.5), rnorm(60, mean = 6.0, sd = 2.5))
)

boxplot(y ~ g, data = d, main = "Look at the data before testing it")
qqnorm(residuals(lm(y ~ g, data = d))); qqline(residuals(lm(y ~ g, data = d)))

print(t.test(y ~ g, data = d))                       # estimate + interval
print(effsize::cohen.d(y ~ g, data = d))             # effect size + CI
print(effsize::cliff.delta(y ~ g, data = d))         # distribution-free

# ---------------------------------------------------------------------
# 2. Bootstrap interval -- no distributional assumption at all.
#    Prefer this to arguing about normality.
# ---------------------------------------------------------------------
diff_means <- function(data, i) {
  s <- data[i, ]
  mean(s$y[s$g == "checklist"]) - mean(s$y[s$g == "unguided"])
}
b <- boot::boot(d, diff_means, R = 5000)
print(boot::boot.ci(b, type = "perc"))

# ---------------------------------------------------------------------
# 3. Nested data -- the error that most inflates significance.
#    Reviews are nested in reviewers and crossed with artefacts.
# ---------------------------------------------------------------------
n_dev <- 34; n_art <- 4
reviews <- expand.grid(dev = factor(1:n_dev), artefact = factor(1:n_art))
reviews$assistant <- rep(c(0, 1), length.out = nrow(reviews))
dev_effect <- rnorm(n_dev, 0, 1.4)[reviews$dev]
art_effect <- rnorm(n_art, 0, 0.8)[reviews$artefact]
reviews$found <- 6 + 0.9 * reviews$assistant + dev_effect + art_effect +
                 rnorm(nrow(reviews), 0, 1.6)

naive <- lm(found ~ assistant, data = reviews)
mixed <- lmerTest::lmer(found ~ assistant + (1 | dev) + (1 | artefact),
                        data = reviews)

cat("\nNaive SE: ", summary(naive)$coefficients["assistant", "Std. Error"],
    "\nMixed SE: ", summary(mixed)$coefficients["assistant", "Std. Error"],
    "\n-- the ratio is how much the naive model overstated precision.\n")
print(confint(mixed, method = "Wald"))

# ---------------------------------------------------------------------
# 4. Counts -- software counts are overdispersed. Poisson gives standard
#    errors that are far too narrow.
# ---------------------------------------------------------------------
cd <- data.frame(size = rpois(300, 40))
cd$defects <- rnbinom(300, mu = 0.05 * cd$size, size = 1.2)   # overdispersed

pois <- glm(defects ~ size, data = cd, family = poisson)
negb <- MASS::glm.nb(defects ~ size, data = cd)

cat("\nPoisson dispersion (should be ~1):",
    sum(residuals(pois, type = "pearson")^2) / pois$df.residual, "\n")
cat("Poisson SE:", summary(pois)$coefficients["size", "Std. Error"],
    " NegBin SE:", summary(negb)$coefficients["size", "Std. Error"], "\n")

# ---------------------------------------------------------------------
# 5. A-priori power. Do this BEFORE collecting, for the smallest effect
#    of interest -- not for the effect you hope to find.
# ---------------------------------------------------------------------
print(pwr::pwr.t.test(d = 0.40, sig.level = 0.05, power = 0.80,
                      type = "two.sample"))

# ---------------------------------------------------------------------
# 6. Equivalence. "No significant difference" is NOT evidence of no
#    difference. Bounds come from your SESOI, not from convention.
# ---------------------------------------------------------------------
print(TOSTER::tsum_TOST(m1 = 6.1, m2 = 5.9, sd1 = 2.4, sd2 = 2.6,
                        n1 = 60, n2 = 60, eqb = 0.40))

# ---------------------------------------------------------------------
# 7. Multiplicity. Holm for confirmatory comparisons declared in the
#    protocol; Benjamini-Hochberg for exploratory work.
# ---------------------------------------------------------------------
pvals <- c(0.011, 0.031, 0.042, 0.180, 0.560)
print(rbind(raw  = pvals,
            holm = p.adjust(pvals, method = "holm"),
            BH   = p.adjust(pvals, method = "BH")))

# ---------------------------------------------------------------------
# 8. Reliability. Check dimensionality FIRST; report omega, not alpha
#    alone -- alpha rises with item count and assumes tau-equivalence.
# ---------------------------------------------------------------------
lat <- rnorm(200)
items <- sapply(1:6, function(i) lat * runif(1, 0.5, 0.8) + rnorm(200, 0, 0.6))
colnames(items) <- paste0("Q", 1:6)

print(psych::fa.parallel(items, fa = "fa", plot = FALSE)$nfact)
print(psych::omega(items, nfactors = 1, plot = FALSE)$omega.tot)
print(psych::alpha(items)$total$raw_alpha)

# ---------------------------------------------------------------------
# 9. Comparing classifiers across datasets. NOT pairwise t-tests:
#    multiplicity is uncorrected and normality is untenable.
# ---------------------------------------------------------------------
res <- matrix(c(0.71,0.69,0.66,0.64,  0.78,0.77,0.74,0.70,
                0.65,0.66,0.61,0.60,  0.82,0.80,0.79,0.74,
                0.69,0.68,0.65,0.63,  0.74,0.73,0.70,0.68),
              nrow = 6, byrow = TRUE,
              dimnames = list(paste0("dataset", 1:6),
                              c("gbm", "rf", "logreg", "size_only")))
print(friedman.test(res))
print(PMCMRplus::frdAllPairsNemenyiTest(res))
cat("\nMean ranks (what the test actually compared):\n")
print(colMeans(t(apply(-res, 1, rank))))

cat("\nAll data here is simulated. See Appendix B of the handout.\n")
